### ISSUE TITLE
Autocomplete not suggesting anything outside of current file

### ISSUE THREAD

**Issue by Muzza** [slynch8/10x#24]
With the latest update, autocomplete seems to suggest things that are not even close to what I'm typing.
For example, if I type 'U', it is suggesting 'gHWSpritesDoubleBuffer', and only that. This variable is declared in the current file, although there are other variables above it that autocomplete is not suggesting.
If there are no variables declared in the file, autocomplete generally does not show up at all.

I'm also getting very inconsistent behaviour across different files. Autocomplete works better in some files than in others and I can't see any obvious reason why. I've been trying to make an easy reproduction case but the behaviour keeps changing and I don't want to spend more time trying to figure it out. 

Here is another weird result it came up with:
![2021-01-07 13_49_38-Window](https://user-images.githubusercontent.com/1583441/103845022-4cc02a00-50ef-11eb-88f5-3647fb9cb7fc.png)
In this case I can see it is simply suggesting local variables before global typedefs (I had to click on 'more' to get UInt16/UInt32 to show up) but the suggestions are clearly way off from what I was typing.

**Comment by slynch8** [comment]
sorry about this, I'll get to it as soon as I can. Until then you might want to revert back to the last version.
https://www.10xeditor.com/prev_versions.htm

**Comment by slynch8** [comment]
in (0.20.1.0) For now I've disabled the "more..." option so it will now show all symbols as before. Can you confirm that this fixes the issue?

**Comment by Muzza** [comment]
It is generally better, thanks. I did find this one cropping up a lot though. It feels odd for it to give all these suggestions (that are not commonly used) ahead of void, which is used everywhere.

![2021-01-08 08_44_30-Window](https://user-images.githubusercontent.com/1583441/103981160-663ba180-51d5-11eb-8c39-8c8817748fdf.png)

**Comment by slynch8** [comment]
oh yes, I still had the C++ keywords under "more...". Fixed in (0.20.5.0). Let me know if you spot anything else.

**Comment by Muzza** [comment]
I usually type 'void(TAB)func' which is still resulting in it changing it to VOID every time.
I've tried accepting the void suggestion (that now appears as the second entry), but it does not prioritise it for the next time I type void.

![2021-01-12 09_35_20-AllProjects_win sln - 10x](https://user-images.githubusercontent.com/1583441/104246128-890dd480-54b9-11eb-9c64-1132dda14b96.png)

**Comment by slynch8** [comment]
in the .workspace_settings (%appdata%\10x\workspaces\) file check the <AutocompleteScores> section. It might be that VOID still has a higher score. You can delete the VOID value and re-open the workspace.

**Comment by Muzza** [comment]
Yes, VOID had a higher score. Removing that entry fixed it.

**Comment by slynch8** [comment]
that's good. I'll add a better interface for this at some point

**Comment by Muzza** [comment]
I noticed it has started doing this again. Every time I type void it suggests something from some 3rd party code. Accepting standard void does not change anything.
I think it started after I deleted all my 10x workspace files.

![2021-03-09 15_35_47-Window](https://user-images.githubusercontent.com/1583441/110419419-667cee00-80ed-11eb-9214-5bd12271cc67.png)

**Comment by slynch8** [comment]
I can't reproduce this issue. Can you check the <AutocompleteScores> section in the workspace file:
%appdata%\10x\Workspaces\<workspace>\<workspace>.workspace_settings

**Comment by Muzza** [comment]
OK, there are a few things going on here, but I have only so much time to test it all.
Autocomplete does not reliably pop up for me (it never has), it will appear on one line, but not another two lines beneath it. I've generally blamed the lack of autocomplete and intellisense on my use of macros which 10x does not fully parse, but I'm really not sure. This makes testing it more confusing.

But here is what I've found:
1. In the workspace file, ALvoid had a score of 3, void a score of 2. I should have remembered this from last time. But I do think it is potentially confusing to users. If A has a score of 10 and AA a score of 1. You have to autocomplete AA 9 times for it to become the preference?

2. From a newly created workspace, it still suggests ALvoid, then ALCvoid first, but only when I do autocomplete within the same namespace as ALvoid and ALCvoid are declared.

3. Typing void then tab will change it to ALvoid, even though void is an exact match for the third item on the autocomplete list. (This is the most annoying thing)

4. If I start to type void outside of the namespace that ALvoid and ALCvoid are declared within, autocomplete suggests VOID first, then void. VOID is defined in a lua c file somewhere. So it is not only a problem within namespaces.

Try this example code, check the comments:

```
namespace NS
{
	typedef void	ALvoid;
	typedef void	ALCvoid;
};

void test1()
{
	// <-- typing void here, the autocomplete works ok
}

namespace NS
{
void test2()
{
	// <-- typing void here, the autocomplete suggests ALvoid/ALCvoid
}
};
```

**Comment by slynch8** [comment]
Thanks for that, I see what's happening now. It sorts the results based on a score, and the score is calculated by various things such as how good a match it is, and how close the result is in scope. Looks like I need to tweak the score weights to prioritise the match score over the scope score. This is one of those things that is really hard to balance correctly in all cases.

If you do find anything that isn't parsing (no autocomplete) that isn't related to the preprocessor then do let me know.

**Comment by slynch8** [comment]
I've tweaked the weights in (0.22.15). Your test case works now, hopefully this works better for you.

**Comment by Muzza** [comment]
Regarding parsing not working, tbh it works in only about 50% of my code, but trying to narrow down the cause of it is not always easy.

For example, I have code like this
```
// header.h
namespace NS1
{
	struct Test1
	{
		int a;
	};

	namespace NS2
	{
		struct Test2
		{
			int a;
		};
	};
};

// file.cpp
#include "header.h"
namespace NS1
{
	using namespace NS2;

	void func()
	{
		Test1 a;	// suggested by autocomplete
		Test2 b;	// not suggested by autocomplete
	}
};
```

In this example Test2 will not be suggested by autocomplete.
If I type b. then it will suggest the member variable from Test2.
In my codebase, Test2 is complete unrecognized by 10x. It has no syntax highlight and won't suggest any member variables or goto the symbol.
I don't know why it partly functions in a stand along example, but does not function at all in my code.

**Comment by Muzza** [comment]
Actually I think I just found one cause. If you put 'using namespace NS2' in a header file rather than the cpp file, it doesn't get parsed.

**Comment by slynch8** [comment]
Yes, you're right that the parser doesn't parse using declarations in headers yet https://github.com/slynch8/10x/issues/16. It doesn't know anything about includes yet, that will go in once I start the preprocessor support.

But your example should work, I'll check that.

**Comment by slynch8** [comment]
autocomplete for 'using namespace' is fixed in (0.22.17)    (for cpp only)

**Comment by slynch8** [comment]
I've opened a new bug for 'using namespace' in header files
https://github.com/slynch8/10x/issues/55
